oauthbearer.go

```diff
@@ -0,0 +1,56 @@
+package sasl
+
+import (
+	"encoding/json"
+	"fmt"
+	"strconv"
+)
+
+// The OAUTHBEARER mechanism name.
+const OAuthBearer = "OAUTHBEARER"
+
+type OAuthBearerError struct {
+	Status  string `json:"status"`
+	Schemes string `json:"schemes"`
+	Scope   string `json:"scope"`
+}
+
+type OAuthBearerOptions struct {
+	Username string
+	Token    string
+	Host     string
+	Port     int
+}
+
+// Implements error
+func (err *OAuthBearerError) Error() string {
+	return fmt.Sprintf("OAUTHBEARER authentication error (%v)", err.Status)
+}
+
+type oauthBearerClient struct {
+	OAuthBearerOptions
+}
+
+func (a *oauthBearerClient) Start() (mech string, ir []byte, err error) {
+	mech = OAuthBearer
+	ir = []byte("n,a=" + a.Username +
+		",\x01host=" + a.Host +
+		"\x01port=" + strconv.Itoa(a.Port) +
```

Comment:
Question: are these fields required? Does Gmail just ignore these?

Reply:
Gmail ignores host and port but fails if a={Username} does not match the token

> Unknown key/value pairs MUST be ignored by the
   server.
https://tools.ietf.org/html/rfc7628#section-3.1

so another question is if there is a case for supporting "any" fields?

Reply:
Okay.

I think we should just add fields to the struct if an RFC extends OAUTHBEARER.